docs/2020-06-22-Rewrite-Rule-Priorities.md

```diff
@@ -0,0 +1,236 @@
+`¬ ∃ X₁ ⌈ C(Y) ∧ φ₁(X₁) ⌉`, so
+applying the second rule when the first one does
+not work is equivalent to computing
+```
+∃ X .
+    ⌈ C(Y) ∧ φ₂(X) ⌉
+    ∧ ¬ ∃ X₁ ⌈ C(Y) ∧ φ₁(X₁) ⌉
+    ∧ ψ₂(X)
+```
+
+Let's see if the two are equivalent:
+
+```
+∃ X .
+    ⌈ C(Y) ∧ φ₂(X) ⌉
+    ∧ ¬ ∃ X₁ . ⌈ C(Y) ∧ φ₁(X₁) ⌉
+    ∧ ψ₂(X)
+===
+// ⌈⌉ commutes with ∃
+∃ X .
+    ⌈ C(Y) ∧ φ₂(X) ⌉
+    ∧ ¬ ⌈ ∃ X₁ . C(Y) ∧ φ₁(X₁) ⌉
+    ∧ ψ₂(X)
+===
+// ∃ X . A ∧ B(X)  ===  A ∧ ∃ X . B(X)
+∃ X .
+    ⌈ C(Y) ∧ φ₂(X) ⌉
+    ∧ ¬ ⌈ C(Y) ∧ ∃ X₁ . φ₁(X₁) ⌉
+    ∧ ψ₂(X)
+===
+// If P is a predicate, ⌈ A ⌉ ∧ P === ⌈ A ∧ P ⌉
+∃ X .
+    ⌈
+        C(Y) ∧ φ₂(X)
+        ∧ ¬ ⌈ C(Y) ∧ ∃ X₁ . φ₁(X₁) ⌉
+    ⌉
+    ∧ ψ₂(X)
+===
+// A ∧ ¬ (A ∧ B) = A ∧ ¬ B
+∃ X .
+    ⌈
+        C(Y) ∧ φ₂(X)
+        ∧ ¬ (C(Y) ∧ ⌈ C(Y) ∧ ∃ X₁ . φ₁(X₁) ⌉)
+    ⌉
+    ∧ ψ₂(X)
+===
+// if A is functional, then A ∧ B = A ∧ ⌈ A ∧ B ⌉
+∃ X .
+    ⌈
+        C(Y) ∧ φ₂(X)
+        ∧ ¬ (C(Y) ∧ ∃ X₁ . φ₁(X₁))
+    ⌉
+    ∧ ψ₂(X)
+===
+// A ∧ ¬ (A ∧ B) = A ∧ ¬ B
+∃ X .
+    ⌈ C(Y) ∧ φ₂(X) ∧ ¬ ∃ X₁ . φ₁(X₁) ⌉
+    ∧ ψ₂(X)
+```
+
+We can generalize this in the following way:
+
+For a rule `φ(X) ⇒ ψ(X) requires P(X) [priority p]` we take the all
+rules at previous priorities:
+```
+φ₁(X) ⇒ ψ₁(X) requires P₁(X)
+...
+φn(X) ⇒ ψn(X) requires Pn(X)
+```
+and we encode the rule as:
+```
+    φ(X) ∧ P(X)
+    ∧ ¬ (∃ X₁ . φ₁(X₁) ∧ P₁(X₁))
+    ...
+    ∧ ¬ (∃ Xn . φn(Xn) ∧ Pn(Xn))
+    ⇒ psi(X)
+```
+Strictly speaking, we should have the left hand side of the encoding of the
+respective rules under `∃ Xᵢ` above, but the two forms should be
+equivalent.
+
+### Priority predicate
+
+If we have only two rules
+```
+φ₁(X) ⇒ ψ₁(X) requires P₁(X) [priority(p₁)]
+φ₂(X) ⇒ ψ₂(X) requires P₂(X) [priority(p₂)]
+```
+and, if we also know that `p₁ < p₂`, we could replace them with
+```
+φ₁(X) ⇒ ψ₁(X) requires P₁(X)
+φ₂(X) ⇒ ψ₂(X)
+    requires P₂(X)
+        ∧ ¬ ∃ X₁ . ⌈φ₂(X) ∧ φ₁(X₁)⌉ ∧ P₁(X₁)
+```
+
+Applying the second rule with a priority predicate means computing:
+```
+∃ X . ⌈ C(Y) ∧ φ₂(X) ∧ P₂(X) ∧ ¬ ∃ X₁ . ⌈φ₂(X) ∧ φ₁(X₁)⌉ ∧ P₁(X₁)⌉ ∧ ψ₂(X)
+===
+// A ∧ ¬ B = A ∧ ¬ (A ∧ B)
+∃ X . ⌈ C(Y) ∧ φ₂(X) ∧ P₂(X) ∧ ¬ ∃ X₁ . φ₂(X) ∧ ⌈φ₂(X) ∧ φ₁(X₁)⌉ ∧ P₁(X₁) ⌉ ∧ ψ₂(X)
+===
+// Assuming φ₂(X) is functional,
+// φ₂(X) ∧ ⌈φ₂(X) ∧ φ₁(X₁)⌉ === φ₂(X) ∧ φ₁(X₁)
+∃ X . ⌈ C(Y) ∧ φ₂(X) ∧ P₂(X) ∧ ¬ ∃ X₁ . φ₂(X) ∧ φ₁(X₁) ∧ P₁(X₁) ⌉ ∧ ψ₂(X)
+===
+// ∃ X₁ . A ∧ B(X)  ===  A ∧ ∃ X₁ . B(X)
+∃ X . ⌈ C(Y) ∧ φ₂(X) ∧ P₂(X) ∧ ¬ (φ₂(X) ∧ ∃ X₁ . φ₁(X₁) ∧ P₁(X₁)) ⌉ ∧ ψ₂(X)
+===
+// A ∧ ¬ (A ∧ B) = A ∧ ¬ B
+∃ X . ⌈ C(Y) ∧ φ₂(X) ∧ P₂(X) ∧ ¬ ∃ X₁ . φ₁(X₁) ∧ P₁(X₁) ⌉ ∧ ψ₂(X)
+===
+// A ∧ ¬ (A ∧ B) = A ∧ ¬ B
+∃ X . ⌈ C(Y) ∧ φ₂(X) ∧ P₂(X) ∧ ¬ ∃ X₁ . C(Y) ∧ φ₁(X₁) ∧ P₁(X₁) ⌉ ∧ ψ₂(X)
+```
+which is the same as above (above `P₁` and `P₂` are included in `φ₁` and `φ₂`).
+
+### Discussion
+
+Using the priority expression (the option that's currently implemented)
+is cleaner form a theoretical point of view since it requires that the
```

Comment:
I'm not sure, but that the significance here is:

```suggestion
is cleaner form a theoretical point of view since it requires only that the
```

Reply:
Thanks, fixed.